### PR TITLE
3/n Simplify spawn plugins: Merge `pre_dispatch` and `setup` logic 

### LINKED ISSUE #10987: [RFC] Revisit/merge/deprecate pre_dispatch, dispatch, post_dispatch and run_stage functions from strategies and trainer
## Proposed refactor

Discussion raised from #10896 

After 2/n simplify spawning logic @awaelchli has done, we can take a step further to simplify hooks

It was in the plan of #10059 in step 4, "deprecate dispatch and post_dispatch"

Will it make sense to deprecate pre/post dispatch and run_stage and aggregate logic into dispatch in both trainer and accelerator?

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @justusschock @awaelchli @akihironitta @kaushikb11 @ananthsub

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of #10987 
Merge after #11149

The concept of "dispatch" no longer exists after #10896. This clean up merges two TTP setup calls that no longer need to be separated. The only minor inconvenience is that the `on_fit_start` call is inbetween the two, but on another note after #10896 the `on_fit_start` hook is no longer _constraint_ to be exactly at this place. 

**Before:**

```python
# trainer.py, _run():

...

self.training_type_plugin.setup(self)  # WE WANT TO MERGE THIS ...
call_hook("on_fit_start")
self.training_type_plugin.pre_dispatch()  #  ... AND THIS

...
```

**After:**

```python
# trainer.py, _run():

...

self.training_type_plugin.setup(self)  # PRE_DISPATCH BECOMES JUST SETUP()
call_hook("on_fit_start")
...
```

Follow-up work will include:
- Renaming restore_checkpoint_after_pre_dispatch
- Move the _log_hyperparams() call to earlier (can now be done after #10896)

### Does your PR introduce any breaking changes? If yes, please list them.

Yes, anyone who implemented `pre_dispatch()` will have to move their code to `setup()`. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

Part of #1 (it's a lie, this is just here to avoid noisy GitHub bot)

cc @tchaton @justusschock @awaelchli @borda

### CODE REVIEW

**Comment by four4fish** [comment]
Note: this will be a breaking change, I saw few use cases user overrides pre_dispatch

File: pytorch_lightning/trainer/trainer.py [hunk 1]
```diff
@@ -1154,8 +1153,9 @@ def _run(
             self._call_callback_hooks("on_fit_start")
             self._call_lightning_module_hook("on_fit_start")
 
-        # plugin will move model to device
-        self._pre_dispatch()
+        # strategy will configure model and move it to the device
+        self.training_type_plugin.setup(self)
```
Comment (four4fish) [inline pytorch_lightning/trainer/trainer.py:1157]: Should the merged setup() happens before call `on_fit_start` hook.

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py:1157]: Yes we will have to figure out if it's possible or not. This branch is WIP and the PR is draft, let me explore all options before I make a proposal

**Comment by ananthsub** [comment]
On naming, users might be confused with the strategy's `setup` running distinctly later than the lightning module, data module, and callbacks `setup`. It'll be especially confusing when we say`setup` is the last hook that runs before checkpointing loading, but then needing to qualify which classes's setups were referring to

**Comment by four4fish** [comment]
> On naming, users might be confused with the strategy's `setup` running distinctly later than the lightning module, data module, and callbacks `setup`. It'll be especially confusing when we say`setup` is the last hook that runs before checkpointing loading, but then needing to qualify which classes's setups were referring to
> 
> Do you think there's another name which captures what's being done in the strategy class here?

@ananthsub that's a really good point. We have setup_distributed and setup_envirment to be the `setup` usually ppl think about. This setup() mainly related to set up optimizer, precision_plugin, and movement logic. I think training_setup()?

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (ananthsub) [reply pytorch_lightning/trainer/trainer.py]: @awaelchli I know we discussed this in the past (but can't find the exact issue/PR) 

with the consolidation around spawning, is this the right spot at all for `on_fit_start`? what prevents this from being inside of the fit loop? 

is `fit` defined as: pretrain routine + sanity check + fit loop ? 

just for the sake of asking this: do we need `on_fit_start` / `on_fit_end` at all if we have the sanity check + on_train_start/end and on_validation_start/end hooks?

Comment (awaelchli) [reply pytorch_lightning/trainer/trainer.py]: Yes, exactly right. Although not a concern for this PR, the discussion of where to place on_fit_start can be restarted again. One can argue different ways:

- on_fit_start can move as early as possible inside the _run() method, but no earlier than setup_environment() since so far it was expected that the process group is initialized in this hook
- for consistency with `on_predict_start`, `on_test_start` etc., on_fit_start should move later into the fit_loop, but the pre-routine/sanity check hooks run earlier than that, making it a confusing choice
- call on_fit_start right when the pre_training_routine starts


>  do we need on_fit_start / on_fit_end at all

It seems like a natural choice to have it. We need to know what people do in this hook before we can decide what to do with it.

File: tests/models/test_hooks.py [hunk 2]
```diff
@@ -516,13 +516,9 @@ def training_step(self, batch, batch_idx):
...
         dict(name="configure_sharded_model"),
         dict(name="Callback.on_configure_sharded_model", args=(trainer, model)),
-        # DeepSpeed skips initializing optimizers here as they are handled via config
-        *([dict(name="configure_optimizers")] if kwargs.get("strategy") != "deepspeed" else []),
+        dict(name="configure_optimizers"),
         dict(name="Callback.on_fit_start", args=(trainer, model)),
         dict(name="on_fit_start"),
-        # TODO: explore whether DeepSpeed can have the same flow for optimizers
-        # DeepSpeed did not find any optimizer in the config so they are loaded here
-        *([dict(name="configure_optimizers")] if kwargs.get("strategy") == "deepspeed" else []),
```
Comment (awaelchli) [inline tests/models/test_hooks.py:525]: Now, the hook order is consistent since there is no longer this problem: 

```py
setup()  # all plugins had optimizers setup here, except deepspeed
call_hook(on_fit_start)
pre_dispatch() # deepspeed had optimizers setup delayed until here
```

Now, setup and pre_dispatch is merged:

```py
setup()
call_hook(on_fit_start)
```

File: pytorch_lightning/plugins/training_type/fully_sharded.py [hunk 3]
```diff
@@ -129,6 +129,14 @@ def setup_distributed(self) -> None:
             )
         super().setup_distributed()
 
+    def setup(self, trainer: "pl.Trainer") -> None:
+        super().setup(trainer)
+        if self.sync_batchnorm:
+            self.model = self.configure_sync_batchnorm(self.model)
+        self.configure_ddp()  # TODO: we call setup_optimizers here too?
+        self.barrier()
```
Comment (awaelchli) [inline pytorch_lightning/plugins/training_type/fully_sharded.py:137]: reminder

File: pytorch_lightning/plugins/training_type/ddp.py [hunk 4]
```diff
@@ -151,6 +151,24 @@ def setup_environment(self) -> None:
...
+        super().setup(trainer)
+        # share ddp pids to all processes
+        self._rank_0_has_called_call_children_scripts = self.broadcast(self._rank_0_has_called_call_children_scripts)
+        if self._should_run_deadlock_detection():
+            self._share_information_to_prevent_deadlock()
+
+        # move the model to the correct device
+        self.model_to_device()
+
+        if self.sync_batchnorm:
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/ddp.py:164]: side question for my knowledge. Why this is still in the codebase if it is not exposed as an argument to the plugin ?

**Review by tchaton — APPROVED** [review]
LGTM !

File: pytorch_lightning/plugins/training_type/ddp.py (hunk 4 above)
Comment (justusschock) [reply pytorch_lightning/plugins/training_type/ddp.py:164]: this is defined by the trainer: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/trainer.py#L167

**Review by carmocca — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/trainer/trainer.py]: Closing this conversation as it is diverging from the PR. Let's move this to an issue.
